### DISCUSSION TITLE
Using Sustain Pedal like Soft Pedal

### DISCUSSION THREAD

**Discussion by RobsonFBP** [FluidSynth/fluidsynth#1271] [Ideas]
### **I propose to use the sustain pedal to increase the release time using a modulator.**

Would be like this:

When the synthesizer reads the modulators and checks for a modulator using source1 == CC-64 and destination volume envelope release, then it would mark the voice to cancel the hold effect.

This way we can have two types of presets with the same instrument:
one instrument would hold the notes until the pedal is released and another would just increase the release time.
With the release time increased, the effect mentioned by @ybungalobill would be achieved.

We would also have guitars, organs and strings that would be tuned.
It will be nice to use piano with pads and strings, like on a real keyboard. (I use a Yamaha PSR-SX900).

This would be similar to using Sustain as a soft pedal without having to use routers, as the modulators are already present in the soundfont.

This is functionality that fluidsynth lacks.
 
**In the SF2 specifications 1.5 Future Enhancements to the SoundFont 2.**
_The SoundFont 2 standard is designed to allow for **enhancements based on future wavetable synthesis
technology** capabilities by additional enumerations of generators and modulators._

Note that this change does not need to be made to the soundfont, as it already has modulators.

It only needs minor synth changes.

How I think this:

1. In the structure of the voice I would add a parameter:
(fluid_voice.h) struct _fluid_voice_t add:
 
`char cancel_default_hold;`

2. Before checking the modulators for the voice set the cancel_default_hold parameter value to false.
(fluid_voice.c) fluid_voice_calculate_runtime_synthesis_parameters: (Add in line 607)

`voice->cancel_default_hold = 0;`

3. While adding the modulators, check if one of them meets the criteria:

4. if it meets the criteria, we mark the cancel_default_hold parameter as true:
(fluid_voice.c) fluid_voice_calculate_runtime_synthesis_parameters (add in line 615 inside the for structure):

      ` if(mod->src1 == SUSTAIN_SWITCH && mod->dest == GEN_VOLENVRELEASE)`
       ` {`
       `      VOICE->cancel_default_hold = 1;`
       ` }`

5. in the noteoff event, we hold the notes only if the cancel_default_hold parameter is false:
(fluid_voice.c) fluid_voice_noteoff check condition for hold the voice (change line 1367)

`    else if(fluid_channel_sustained(channel) && voice->cancel_default_hold == 0)`
      `    {`
          `        voice->status = FLUID_VOICE_SUSTAINED;`
     `     }`
       

Dear @derselbst, please check these proposed changes.
Thanks.

**Comment by derselbst** [comment]
Well, I'm not really convinced. The original use-case in #1214 was described as:

> On a real piano, if one releases a key and then quickly depresses the sustain pedal, the remaining vibrations from that note will be sustained

So essentially what happens in your proposal is that `else if(fluid_channel_sustained(channel)` will be processed at the time of the noteOff, i.e. when the channel is not yet sustained, hence the voice will enter release phase. There is no logic in your solution that would make the voice become "unreleased" when the channel becomes sustained right afterwards. (This is important though, because a soundfont designer may choose to set "sustain level" for an instrument where the instrument would stay at a certain amplitude, whereas in the release phase there'll always be a linear decrease in volume.)

Instead, the modulator would indeed result in a sudden increase of the release time whenever CC64 becomes >64, highly depending on the amount that was given for that modulator, which you have left unspecified. So if that's the solution you're looking for, you're free to specify those modulators.

The need for this `cancel_default_hold` variable is questionable. If you wanted to control sustain and release-time independently / exclusively from each other, simply using a different source CC for that release-time-modulator would be sufficient.

**Comment by RobsonFBP** [comment]
Hello @derselbst 
> There is no logic in your solution that would make the voice become "unreleased" when the channel becomes sustained right afterwards.

This doesn't seem to make sense, but...
Let's try it with a graphical example.
We imagine our release time to be 0.3s.

![normal volume envelope](https://github.com/FluidSynth/fluidsynth/assets/142550370/5dc577be-cd47-4b62-987e-7029c570f0ee)

When the pedal is depressed, the modulator adds 4s to the release time.
If I depress the pedal in the release phase, the release time increases, resulting in the desired effect mentioned in #1214  by @ybungalobill.

![normal volume envelope3](https://github.com/FluidSynth/fluidsynth/assets/142550370/e69cafad-46f6-4a0a-aab6-89fca5ffe6dc)

> Instead, the modulator would indeed result in a sudden increase of the release time whenever CC64 becomes >64, highly depending on the amount that was given for that modulator, which you have left unspecified.

I didn't specify the amont value because each instrument can use different values. In this example amont value is 4800.

![modulador CC-64 dest vol env release](https://github.com/FluidSynth/fluidsynth/assets/142550370/933708c6-9172-47f8-9765-d27520aefe55)

>The need for this `cancel_default_hold` variable is questionable. If you wanted to control sustain and release-time independently / exclusively from each other, simply using a different source CC for that release-time-modulator would be sufficient.

The idea is to use the pedal to increase the release time using the modulator. 

While the notes are in sustain, the effect of increasing the release time does not occur. If I release the pedal, the voice enters the release phase, but the modulator also returns the release time to the initial phase, in our example 0.3 s. That's why I suggested creating the `cancel_default_hold` variable.

In addition to the expected effect, we have doubled the creative ability to create sounfonts.
Sounds like dunes, organs and strings can now sound like real keyboards.
We could have two equal presets, one using this effect and the other not using it.

It would be grand!

**Reply by derselbst** [reply]
Ok. If I find some time in the next days / weeks, I'll set up a PR for this and we'll how / if it works. I really don't know what happens when we manipulate fluidsynth volume env on the fly... never tried this before.

**Reply by RobsonFBP** [reply]
Thanks.

**Comment by RobsonFBP** [comment]
### **_SUSTAIN STUFS_**

**Hello @ybungalobill,  @derselbst, @cghawthorne, @atsushieno** 

I still think we should continue trying to achieve the result with the sustain pedal.
I propose a new way to achieve this.

The idea is to increase the release time using the sustain pedal, making the voice have a smooth fall, a kind of fadeout, so that it doesn't sound indefinitely.

In some situations, we can also use the sustain pedal to function as a sostenuto.

1. Let's define a _sustain_mode enumerator in fluid_voice.h
`   enum _sustain_mode `
`{                                    `
`  NOTE_OFF_PENDED,   `
`  IGNORE_SUSTAIN,       `
`  MODE_SOSTENUTO    `
`};                                   `

2. we create an unsigned char sustain_mode variable
 inseert in fluid_voice.h linha 71 in struct _fluid_voice_t in fluid_voice.c

	`/* sustain_mode from special criteria modulator */`
`	unsigned char sustain_mode; /* see enum _sustain_mode */`

3. We initialize the assigned voice to the default value for the sustain_mode variable

     insert in new_fluid_voice at line 238 in fluid_voice.c
	`voice->sustain_mode = NOTE_OFF_PENDED;`

5. If any modulator meets the criteria, we modify the value of the sustain_mode variable
	
	insert in fluid_voice_calculate_runtime_synthesis_parameters in fluid_voice.c
	at line 616 inside for structure.

`  if(voice->sustain_mode == NOTE_OFF_PENDED)`
`  {`
`	if(mod->src1 == SUSTAIN_SWITCH && `
`			mod->src2 == HOLD2_SWITCH && `
`			mod->dest == GEN_VOLENVRELEASE)`
`	{`
`		voice->sustain_mode = IGNORE_SUSTAIN;`
`	}`
`	else if(mod->src1 == SUSTAIN_SWITCH && `
`			mod->src2 == SOSTENUTO_SWITCH && `
`			mod->amount == 0)`
`	{`
`		voice->sustain_mode = MODE_SOSTENUTO;`
`	}`
`  }`


### **NOTE:**
We do not check the flags because the index values ​​(64, 65, 66, 67,68, 69) do not have correspondences in the General Controller, therefore, these will be Continuous Controller.

see SF Specs 8.2.1 Source Enumerator Controller Palettes.

As a stuck crit I suggest using:

A. source1 = CC64 (Sustain switch) and source2 = CC69 (Hold2_switch) and target VolEnvRelease to ignore sustain

![aa](https://github.com/FluidSynth/fluidsynth/assets/142550370/47b76994-5071-4e6f-8d11-4a967b5eeaf1)

B. source1 = CC64 (Sustain switch) and source2 = CC66 (Sostenuto_switch) and amount = 0 to function as sostenuto

![ba](https://github.com/FluidSynth/fluidsynth/assets/142550370/d649c5da-f8bd-4b48-bd97-e32aed9e4bc5)


Logically, the soundfont programmer should use direction 1 in the second source:

6. When releasing the note with the sustain pedal pressed, it produces the desired effect.

insert influid_voice.c void fluid_voice_noteoff(fluid_voice_t *voice)

`     /* Or sustain a note under Sustain pedal */`
`    else if(fluid_channel_sustained(channel) && `
`			(voice->sustain_mode == NOTE_OFF_PENDED ||`
`            (channel->sostenuto_orderid > voice->id &&`
`			voice->sustain_mode == MODE_SOSTENUTO)))`
`	{`
`           voice->status = FLUID_VOICE_SUSTAINED;`
`       }      `
